### ISSUE TITLE
To test at build? Or not to test?

### ISSUE THREAD

**Issue by biznickman** [mxstbr/react-boilerplate#591]
This is a continuation of the conversation we started here https://github.com/mxstbr/react-boilerplate/issues/561 ... The question is whether or not we want to move the testing out of the build process. Specifically the `prebuild` task in package.json can result in slowed deployments on Heroku or any other deployment location where the build takes on the server.

**Comment by hamishtaplin** [comment]
I've had to do this for the very reason you've specified, I added an extra `compile` npm task:

```"scripts": {
    ...,
    "prebuild": "npm run test && npm run build:clean",
    "build:clean": "rimraf ./qBuilder/bundles",
    "build": "npm run compile",
    "compile": "cross-env NODE_ENV=production webpack --config internals/webpack/webpack.prod.babel.js --color -p"```

**Comment by gaboesquivel** [comment]
IMHO build shouldn't call test. so you can build without testing if you're not interested in that for any reason. You can always run test and then build.   `test && build` 
deployment task should only take the `build` dir and push it to the hosting service.  

in sum:  all tasks independent so you have more control  `test && build && deploy` or just `build && deploy`

**Comment by biznickman** [comment]
@gaboesquivel I agree. If you were using something like continuous integration, the tests would be done as well prior to deployment. I think eliminating test from the build process makes the most sense. It's attempting to accomplish something that should essentially be managed by deployment workflow rather than application-level workflow. I'm curious to see if there is any argument in support of it.

**Comment by datapimp** [comment]
I've personally started separating out the main component of each of the different test processes so that I can call them without testing / linting when I want, using something like `npm run test:only` 

i like that building forces me to run the linter and the tests; but i also hate it.  I think it should be a goal to keep the abilities extremely discoverable and accessible but also make them easy to turn on and off without modifying a bunch of files and remembering dependencies to install;  

This is one area that I think the boilerplate having its own CLI can really clean up the UX, and hide some of the scripts which may be more important for the development of the  react-boilerplate project than they are for the development of every project based on the boilerplate;  

For example:

`react-boilerplate build --require-linter --require-tests` 

I think we want this to be the case for master branch commits running in CI especially on the boilerplate project itself;  

I don't think this is a helpful default for live projects running in production.  Most of the time I'm not gonna have the chance to deploy to production without extreme confidence these steps have been run against the particular git SHA so it is always a pain.

I think it is much easier to reason about CLI arguments than it is to decipher the npm package.json and lifecycle hooks system. I'm working on another project with the intention of making accidental progress on a boilerplate cli; with hopes of submitting a PR eventually.

I would love to see somebody run with this though.

**Comment by gihrig** [comment]
> I think it is much easier to reason about CLI arguments than it is to decipher the npm package.json and lifecycle hooks system.

I agree, it is desirable to simplify and clean up package.json.

What is being asked for in this issue is a simple way to bypass testing when, for whatever reason a production build is desired on a project that will not pass test. This will be more common now that test coverage of 98%+ is a requirement for the testing phase to pass.

This can be changed simply without altering existing workflow.

If there are no objections, I'll open a PR.

Thoughts?

**Comment by mxstbr** [comment]
I think a PR would be very welcome for this @gihrig!

**Comment by gihrig** [comment]
Done at #1432, merged and included in 3.4 😁 

Closing
